producer/data/ref_counted.go

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package data
+
+import (
+	"sync"
+
+	"github.com/m3db/m3msg/producer"
+
+	"go.uber.org/atomic"
+)
+
+// OnFinalizeFn will be called when the data is being finalized.
+type OnFinalizeFn func(d producer.RefCountedData)
+
+type refCountedData struct {
+	sync.RWMutex
+
+	rc       *atomic.Int32
+	d        producer.Data
+	fn       OnFinalizeFn
+	isClosed bool
+}
+
+// NewRefCountedData creates RefCountedData.
+func NewRefCountedData(data producer.Data, fn OnFinalizeFn) producer.RefCountedData {
+	return &refCountedData{
+		rc:       atomic.NewInt32(0),
+		d:        data,
+		fn:       fn,
+		isClosed: false,
+	}
+}
+
+func (d *refCountedData) Filter(fn producer.FilterFunc) bool {
+	return fn(d.d)
+}
+
+func (d *refCountedData) IncRef() {
+	d.rc.Inc()
+}
+
+func (d *refCountedData) DecRef() {
+	rc := d.rc.Dec()
+	if rc == 0 {
+		d.finalize(producer.Consumed)
+	}
+	if rc < 0 {
+		panic("invalid ref count")
+	}
+}
+
+func (d *refCountedData) Bytes() ([]byte, bool, producer.DoneFn) {
+	d.RLock()
```

Comment:
because while a writer is using the byte, aka writing the bytes out, the data could be dropped by the buffer, so it needs a read lock while the bytes are being used by the writer to avoid the bytes being recycled.

Reply:
Hm I wonder if we could design this better to avoid returning while holding the lock, seems unconventional.

Reply:
it kinda needs a lock because the byte slice will be concurrently accessed by multiple threads, so I need access control + blocking/wait for the bytes.

Reply:
Can you pass in a callback and call that callback within a read lock instead?

Reply:
I gave that a try, but that would lead to much less efficient proto.Message using, since I need to get a new message from the pool every time I retry writing it out rather than just holding on to one proto.Message for that data and only call this function to just get the bytes before encoding. Also that would require me to create a new lambda to pass here per write since I need to capture a different metadata for shard and id in the lambda. Both seemed less efficient.

Reply:
Changed the interface to expose IncReads/DecReads instead of DoneFn for this use case.